elastic/logstash
# metrics: remove delayed implementation of timers interface

## Release notes

 - Improves the timer metrics that measure execution and queue operations. These metrics now include in-flight timings that have not yet completed so that blockages can be detected before they are resolved, and they accumulate "leftover" fractional milliseconds across timings to improve accuracy.

_NOTE: the above release notes cover the combined work done in #14748 **and** this PR._

## What does this PR do?

The delayed implementation `AfterCompletionTimerMetric` of the `TimerMetric` interface, introduced along-side that interface's introduction to replicate the previous (undesired) behaviour, is superceded by an already-merged live-tracking implementation that is effectively as performant when not under concurrent contention and still reasonably performant when a single timer is contended across multiple threads.

The `metric.timers` setting removed here has not been a part of any Logstash release and can safely be removed without going through the normal deprecation path; from the user's perspective this removal combined with the previously-merged work is simply an improvement to the accuracy of the existing timer metrics exposed via our API.

## Why is it important/What is the impact to the user?

As described in https://github.com/elastic/logstash/issues/14675 the old way of reporting timings after completion lead to metrics becoming _least_ useful when they were needed _most_:

> For example, a grok filter handling input that is reliably causing it to time-out with the default 30s per match, handling the default batch size of 125, can go ~62.5 _minutes_ without incrementing the relevant metrics about its execution time. If all of a pipeline's workers are in this state, its `worker_concurrency`'s `current` flow rate can report 0, _despite all workers being fully-occupied_.
> 
> Similarly, a pipeline whose inputs are _entirely_ blocked pushing events into the queue will not report time spent in _queue-push-wait_ until _after_ the batches that are blocked have been accepted (which for a fully-stalled pipeline may be _never_). If all of a pipelines inputs are entirely blocked, its `queue_backpressure`' s `current` flow rate can report a current rate of 0, _despite all inputs being fully-blocked_.

After delivering the API boundary for timers as a first-class metric, a temporary net-no-change implementation, and a new live-tracking implementation, we have determined that the live-tracking is a suitable drop-in replacement and that the temporary implementation can be removed.

## Checklist




## How to test this PR locally

Run Logstash with a pipeline (the one below runs at ~1000 events/sec/worker in non-prime minute-of-the-hour and much slower in prime minutes; when run with >=3 workers it catches up and falls behind):

~~~
input { java_generator { threads => 2 eps => 1000 } }
filter {
  ruby {
    id => "slow_prime"
    init => 'require "prime"'
    code => "sleep((Prime.prime?(Time.now.min) && (Random.rand(100) <= 1)) ? (2.0 ** Random.rand(4.0)) : 0.001)"
  }
}
output { sink {} }
~~~

Observe the node stats API's flow metrics like `worker_concurrency` and `queue_backpressure` are non-zero during prime minutes when all workers are occupied doing long-running work and all inputs are blocked on queue push operations. This is because their denominator (a timer metric) is incrementing even when those operations are blocked.

```
 watch "curl -XGET 'localhost:9600/_node/stats' | jq .pipelines.main.flow"
```

## Related issues

- Closes #14809
